### DISCUSSION TITLE
AssertionError: vivado_hls not found in PATH

### DISCUSSION THREAD

**Discussion by caoimhelynch** [Xilinx/finn#563] [Q&A]
I am attempting to run cnv_end2end_example.ipynb. I am at the hardware generation stage. I am getting the error AssertionError: vivado_hls not found in PATH. 

I have the path C:\Xilinx\Vivado\2020.1\bin\unwrapped\win64.o set up in the PATH environment variables. I also have C:\Xilinx\Vivado\2020.1 and C:\Xilinx set as requested by the FINN requirements. From debugging the file is checked, however basic.py on line 496 defines the file path as not executable with the function 'is_exe(exe_file)' where C:\Xilinx\Vivado\2020.1\bin\unwrapped\win64.o is passed through as exe_file. I am unsure why as there is a vivado_hls file in this location however it does not seem to be recognised by this function. 

I am wondering is there an issue with the way I have vivado installed that this is the only file that contains vivado_hls? Should a different executable file be available somewhere that I am missing? I have 2020.1 version installed as required by the FINN getting started page. 

Appreciate any help.

**Comment by fpjentzsch** [comment]
Hi,
are you trying to use FINN on Windows? Vivado should be installed on the same Linux OS (preferably Ubuntu 18.04) as FINN.

**Reply by caoimhelynch** [reply]
Hi thanks for getting back to me. Yes I am on Windows, I have Ubuntu 18.04 installed however FINN and vivado I believe are both installed locally. How do I install these on Ubunto for use? 

I am not using Jupyter notebooks either I am running my code from a PyCharm IDE just because I was more familiar with this.

**Reply by fpjentzsch** [reply]
Is your Ubuntu 18.04 installed in a VM or are you using the Windows Subsystem for Linux (WSL)?

**Reply by caoimhelynch** [reply]
WSL I believe, I have installed it as a Windows application

**Reply by fpjentzsch** [reply]
If you use WSL 2 (see this guide to check/install: https://docs.microsoft.com/en-us/windows/wsl/install), it is possible to install Docker, FINN, and Vivado within it, as it is very similar to a VM under the hood. You will have no desktop environment (at least not without a lot of tinkering), so you'll have to install the Xilinx tools via command line. You won't be able to invoke the Vivado/HLS GUI, which is no problem for FINN internally, just if you want to manually inspect a project created by FINN. (That is, unless you are on Windows 11 and make use of the new WSLg feature to run GUI applications from within WSL 2... but I have not tested this myself yet)

The easier option would probably be to run Ubuntu natively or in a "real" VM with a desktop environment.

**Reply by caoimhelynch** [reply]
I believe I will need a desktop environment for Vivado to see the component cost on FPGA, this is the aim for what I am doing. So in this case you are saying Ubuntu natively or on a VM is the best option? Would you be able to point me in the right direction for documentation on how to do this? There seemed to be many ways to install Ubuntu and I have never used it so was quite confused. 

Will installing this natively mean I can run my code the same way I am now, and installing on a VM mean I have to run my code on that VM? Does installing 'natively' involve a flash drive? Apologies if these are novice questions I am not at all familiar with Ubunto or Linux environments in general.

**Reply by caoimhelynch** [reply]
Hi @fpjentzsch I just want to clarify before I go forward setting up a Ubuntu vm, as I am under time pressure for this University assignment. 

Following the endtoend example: 
https://github.com/Xilinx/finn/blob/d1cc9cf94f1c33354cc169c5a6517314d0e94e3b/notebooks/end2end_example/bnn-pynq/tfc_end2end_example.ipynb 
I am only attempting to reach the line of input 103 in the section 'Examining the generated outputs.' 
This is to see the vivado project and get the cost implementation of my model. This is as far as I intend on going through the process. 

To get to this point is there any way around the path issue or do I definitely need a VM with Ubuntu to achieve this? Really appreciate if you could clarify for me.

**Reply by fpjentzsch** [reply]
To see the final hardware cost/FPGA utilization, you'll need to run Vivado HLS synthesis and Vivado synthesis. You don't need a GUI to view the respective report, the model metadata property `vivado_synth_rpt` should point to an .xml file that lists hierarchical resource usage.

I never tried to invoke a Windows Vivado install from within FINN. Maybe there is a way to make it work, but I can't help you with that. Theoretically, you could take whatever code/scripts/commands FINN generates and passes to Vivado HLS/Vivado internally, and then run them manually with your Windows-installed Vivado. However, you'll run into a slew of new issues, such as dependencies between IP cores and Windows file path length limitations.

Yes, with "natively" I mean that you boot from an install drive and install a fresh Ubuntu OS on some computer. I'm not entirely sure what your current setup with  PyCharm looks like, but FINN is intended to run only inside the provided docker container (which itself runs under Linux or the WSL 2 if you are on Windows and use Docker Desktop), using either the integrated Jupyter notebook server, the "build_custom" build flow feature, or simply the "bash" shell.

I really recommend to get familiar with Linux. If you don't like the idea of working within a separate (and potentially slow) VM, maybe WSL 2 is the right way for you. Using Visual Studio Code, you can connect to a WSL 2 workspace and edit/execute your code from there as you normally would, except that it runs under Linux/Ubuntu.

**Reply by caoimhelynch** [reply]
Okay I really appreciate you getting back to me. To clarify I can use the WSL along with Visual Studio code to execute the code under Ubuntu and I would not see any GUI but could still get the .xml file that will display the costs/resource usage? Or I can attempt to continue with windows but if I run into any issues I cannot get any help on this, and you suspect there will be issues? 

I have managed to get past the error I was having and I am now running into issues of  

Exception: Custom op_type IODMA is currently not supported.

and 

Traceback (most recent call last):
  File "C:\Users\Caoimhe\PycharmProjects\Computational-Imagaing-FPGA-gitlab\finn\src\finn\transformation\fpgadataflow\hlssynth_ip.py", line 69, in applyNodeLocal
    inst.ipgen_singlenode_code()
  File "C:\Users\Caoimhe\PycharmProjects\Computational-Imagaing-FPGA-gitlab\finn\src\finn\custom_op\fpgadataflow\hlscustomop.py", line 323, in ipgen_singlenode_code
    builder.build(code_gen_dir)
  File "C:\Users\Caoimhe\PycharmProjects\Computational-Imagaing-FPGA-gitlab\finnbase\src\finn\util\hls.py", line 65, in build
    working_dir = os.environ["PWD"]
  File "C:\Users\Caoimhe\anaconda3\lib\os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'PWD'

There doesn't appear to be any documentation on this IODMA error at all and it is not clear why my op_type is IODMA or how I can fix this. Have you any advice on this? Thanks

**Reply by fpjentzsch** [reply]
Well, the first error (IODMA) likely occurs because you are trying to run FINN code in a regular IDE on Windows? FINN has complex dependencies, that's why we always run FINN code inside the docker container (using the "run-docker.sh" script). The second error occurs because you are running it on Windows, but FINN only supports Linux.

In general you can use WSL & VS Code that way, yes. However, because FINN code needs to run in this container, you would not execute the python code directly inside your WSL 2 workspace in VS Code (editing is fine). You'd have to use the terminal (it's also integrated into VS Code) to start "run-docker.sh". If you setup Docker Desktop correctly, you could then connect VS Code to a workspace within this container as well. I suppose in that scenario you could run your code from within VS Code like you seem to expect.
